File: wolf_nlp/alg_practice/aho-corasick/aho-corasick.cpp

#include <cstdlib>
#include <set>
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

#define ALPHABET_NUMBER 26

struct StateNode
{
    bool finish_{ false };
    int state_{ 0 };
    string pattern_{};
    //goto table
    vector<StateNode *> transition_table_{ vector<StateNode *>(ALPHABET_NUMBER) };
};

class ACSM
{
private:
    StateNode *start_node_;
    int state_count_;
    vector<StateNode *> corresponding_node_;
    vector<StateNode *> fail_;
public:
    ACSM() :start_node_{ new StateNode() }, state_count_{ 0 }
    {
        //state0 is start_node_
        corresponding_node_.push_back(start_node_);
    }
    //read all patterns and produce the goto table
    void load_pattern(const vector<string> &_Patterns)
    {
        int latest_state = 1;
        for (const auto &pattern : _Patterns)
        {
            auto *p = start_node_;
            for (int i = 0; i < pattern.size(); ++i)
            {
                auto *next_node = p->transition_table_[pattern[i] - 'a'];
                if (next_node == nullptr)
                {
                    next_node = new StateNode();
                }
                if (next_node->state_ == 0)
                {
                    next_node->state_ = latest_state++;
                    //update the table
                    corresponding_node_.push_back(next_node);
                }
                //the goto table
                p->transition_table_[pattern[i] - 'a'] = next_node;
                p = next_node;
            }
            p->finish_ = true;
            p->pattern_ = pattern;
        }
        for (int i = 0; i < ALPHABET_NUMBER; ++i)
        {
            if (start_node_->transition_table_[i] == nullptr)
            {
                start_node_->transition_table_[i] = start_node_;
            }
        }
        state_count_ = latest_state;
    }
    //produce fail function
    void dispose()
    {
        queue<StateNode *> q;
        fail_ = std::move(vector<StateNode *>(state_count_));
        // 给深度为1的复制为0
        for (const auto nxt : start_node_->transition_table_)
        {
            //d=1,f=0
            if (nxt->state_ != 0)
            {
                fail_[nxt->state_] = start_node_;
                q.push(nxt);
            }
        }
        //calculate all fail redirection
        while (!q.empty())
        {
            auto known = q.front();
            q.pop();
            for (int i = 0; i < ALPHABET_NUMBER; ++i)
            {
                auto nxt = known->transition_table_[i];
                if (nxt && nxt->state_ != 0)
                {
                    //上一级的fail的同条件的，
                    //比如：如果在状态8失败，那么先求fail(7)，然后在fail(7)条件下goto，条件为到从7到8的条件c，然后goto(fail(7),c)
                    auto p = fail_[known->state_];
                    //这里...？？？
                    while (!p->transition_table_[i])
                    {
                        p = fail_[p->state_];
                    }
                    fail_[nxt->state_] = p->transition_table_[i];
                    q.push(nxt);
                }
            }
        }
    }
    //search matching
    void match(const string &_Str, set<string> &_S)
    {
        auto p = start_node_;
        for (int i = 0; i < _Str.size(); ++i)
        {
            int trans = _Str[i] - 'a';
            p = p->transition_table_[trans] ? p->transition_table_[trans] : (--i, fail_[p->state_]);
            if (p->finish_)
            {
                _S.insert(p->pattern_);
            }
        }
    }
};

int main()
{
    ACSM acsm;
    vector<string> patterns{ "his","hers","she","he" };
    set<string> matched;
    acsm.load_pattern(patterns);
    acsm.dispose();
    string str{ "hishers" };
    acsm.match(str, matched);
    for (const auto str : matched)cout << str << endl;
    system("pause");
    return 0;
}

File: wolf_nlp/alg_practice/aho-corasick/aho-corasick.py
#coding=utf8
import queue

class StateNode(object):
    def __init__(self, finish=False, state=0, pattern=None, transition_table=None):
        self.finish_ = finish
        self.state_ = state
        self.pattern_ = pattern
        self.transition_table_ = transition_table if transition_table else dict()

class AC(object):
    def __init__(self):
        self.start_node = StateNode()
        self.state_count = 0
        self.correctponding_node = []
        self.fail = {}
        self.alphabet_num = 26

    def load(self, patterns):
        latest_state = 1
        for item in patterns:
            p = self.start_node
            for i in item:
                next_node = p.transition_table_.get(ord(i)-ord('a'), None)
                if not next_node:
                    next_node = StateNode()
                if next_node.state_ == 0:
                    next_node.state_ = latest_state
                    latest_state += 1
                    self.correctponding_node.append(next_node)
                p.transition_table_[ord(i) - ord('a')] = next_node
                p = next_node
            p.finish_ = True
            p.pattern_ = item

        self.state_count = latest_state
        for i in range(self.alphabet_num):
            if not self.start_node.transition_table_.get(i, None):
                self.start_node.transition_table_[i] = self.start_node

        self.dispose()

    def dispose(self):
        q = queue.Queue()
        self.fail = [StateNode() for _ in range(self.state_count)]
        for item in self.start_node.transition_table_.values():
            if item.state_ != 0:
                self.fail[item.state_] = self.start_node
                q.put(item)

        while not q.empty():
            known = q.get()
            for i in range(self.alphabet_num):
                nxt = known.transition_table_.get(i, None)
                if nxt and nxt.state_ != 0:
                    p = self.fail[known.state_]
                    while not p.transition_table_[i]:
                        p = self.fail[p.state_]
                    self.fail[nxt.state_] = p.transition_table_[i]
                    q.put(nxt)

    def match(self, s):
        res = []
        p = self.start_node
        i = 0
        while i < len(s):
            if p.transition_table_.get(ord(s[i])-ord('a'), None):
                p = p.transition_table_.get(ord(s[i])-ord('a'), None)
            else:
                p = self.fail[p.state_]
                i -= 1
            if p.finish_:
                res.append(p.pattern_)
            i += 1
        return res

if __name__ == '__main__':
    ac = AC()
    pts = []
    pts.append('hers')
    pts.append('she')
    pts.append('his')
    ac.load(pts)
    s = 'asldflsdhflkhessldhflasdhfshesdlfhsakdhishsdhflaskdhfksad'
    res = ac.match(s)
    print(res)
